### DISCUSSION TITLE
paketo buildpacks via buildbootimage vs pack cli

### DISCUSSION THREAD

**Discussion by Rvarman0110** [paketo-buildpacks/feedback#276] [Java Team]
In my root directory I have 

**gradle.properties
gradlew
gradlew.bat
settings.gradle**

When I do **_./gradlew buildbootimage_**

These things are getting participated 

[creator]     paketo-buildpacks/ca-certificates   3.6.8
    [creator]     paketo-buildpacks/bellsoft-liberica 10.5.5
    [creator]     paketo-buildpacks/syft              1.45.0
    [creator]     paketo-buildpacks/executable-jar    6.8.5
    [creator]     paketo-buildpacks/dist-zip          5.6.10
    [creator]     paketo-buildpacks/spring-boot       5.27.11
    
    and it is setting up a process type 
    
    [creator]       **Process types:**
    [creator]         executable-jar: java org.springframework.boot.loader.launch.JarLauncher (direct)
    [creator]         task:           java org.springframework.boot.loader.launch.JarLauncher (direct)
    [creator]         web:            java org.springframework.boot.loader.launch.JarLauncher (direct)
    [creator]
    [creator]     Paketo Buildpack for Spring Boot 5.27.11
    [creator]       https://github.com/paketo-buildpacks/spring-boot
    [creator]       Build Configuration:
    [creator]         $BP_SPRING_CLOUD_BINDINGS_DISABLED   false  whether to contribute Spring Boot cloud bindings support
    [creator]         $BP_SPRING_CLOUD_BINDINGS_VERSION    1      default version of Spring Cloud Bindings library to contribute
    [creator]       Launch Configuration:
    [creator]         $BPL_SPRING_CLOUD_BINDINGS_DISABLED  false  whether to auto-configure Spring Boot environment properties from bindings
    [creator]         $BPL_SPRING_CLOUD_BINDINGS_ENABLED   true   Deprecated - whether to auto-configure Spring Boot environment properties from bindings
    [creator]       Creating slices from layers index
    [creator]         dependencies (122.5 MB)
    [creator]         spring-boot-loader (446.0 KB)
    [creator]         snapshot-dependencies (0.0 B)
    [creator]         application (264.3 KB)
    [creator]       Launch Helper: Contributing to layer
    [creator]         Creating /layers/paketo-buildpacks_spring-boot/helper/exec.d/spring-cloud-bindings
    [creator]       Spring Cloud Bindings 2.0.3: Contributing to layer
    [creator]         Downloading from https://repo1.maven.org/maven2/org/springframework/cloud/spring-cloud-bindings/2.0.3/spring-cloud-bindings-2.0.3.jar
    [creator]         Verifying checksum
    [creator]         Copying to /layers/paketo-buildpacks_spring-boot/spring-cloud-bindings
    [creator]       Web Application Type: Contributing to layer
    [creator]         Servlet web application detected
    [creator]         Writing env.launch/BPL_JVM_THREAD_COUNT.default
    [creator]       4 application slices
    [creator]       Image labels:
    [creator]         org.opencontainers.image.title
    [creator]         org.opencontainers.image.version
    [creator]         org.springframework.boot.version
    
    
    
    

but while using the same builder in pack command
    
    these buildpacks are getting detected 
    11 of 26 buildpacks participating
paketo-buildpacks/ca-certificates   3.6.8
paketo-buildpacks/bellsoft-liberica 10.5.5
paketo-buildpacks/syft              1.45.0
paketo-buildpacks/gradle            7.9.0
paketo-buildpacks/executable-jar    6.8.5
paketo-buildpacks/apache-tomcat     7.15.3
paketo-buildpacks/apache-tomee      1.8.2
paketo-buildpacks/liberty           4.0.4
paketo-buildpacks/dist-zip          5.6.10
paketo-buildpacks/spring-boot       5.27.11

Its executing the java and gradle buildpack then its not executing any spring boot and it is also not setting a process type why ? 

logs:
[builder] Paketo Buildpack for Syft 1.32.1
[builder]   https://github.com/paketo-buildpacks/syft
[builder]     Downloading from https://github.com/anchore/syft/releases/download/v0.84.0/syft_0.84.0_linux_amd64.tar.gz
[builder]     Verifying checksum
[builder]     Writing env.build/SYFT_CHECK_FOR_APP_UPDATE.default
[builder]
[builder] Paketo Buildpack for Gradle 7.4.0
[builder]   https://github.com/paketo-buildpacks/gradle
[builder]   Build Configuration:
[builder]     $BP_EXCLUDE_FILES                                                                       colon separated list of glob patterns, matched source files are removed
[builder]     $BP_GRADLE_ADDITIONAL_BUILD_ARGUMENTS                                                   the additionnal arguments (appended to BP_GRADLE_BUILD_ARGUMENTS) to pass to Gradle
[builder]     $BP_GRADLE_BUILD_ARGUMENTS             --no-daemon -Dorg.gradle.welcome=never assemble  the arguments to pass to Gradle
[builder]     $BP_GRADLE_BUILD_FILE                                                                   the location of the main build config file, relative to the application root
[builder]     $BP_GRADLE_BUILT_ARTIFACT              build/libs/*.[jw]ar                              the built application artifact explicitly.  Supersedes $BP_GRADLE_BUILT_MODULE
[builder]     $BP_GRADLE_BUILT_MODULE                                                                 the module to find application artifact in
[builder]     $BP_GRADLE_INIT_SCRIPT_PATH                                                             the path to a Gradle init script file
[builder]     $BP_INCLUDE_FILES                                                                       colon separated list of glob patterns, matched source files are included
[builder]     $BP_JAVA_INSTALL_NODE                  false                                            whether to install Yarn/Node binaries based on the presence of a package.json or yarn.lock file
[builder]     $BP_NODE_PROJECT_PATH                                                                   configure a project subdirectory to look for `package.json` and `yarn.lock` files
[builder]     Creating cache directory /home/cnb/.gradle
[builder]   Compiled Application: Contributing to layer
[builder]     Executing gradlew --no-daemon -Dorg.gradle.welcome=never assemble
[builder]       To honour the JVM settings for this build a single-use Daemon process will be forked. For more on this, please refer to https://docs.gradle.org/8.7/userguide/gradle_daemon.html#sec:disabling_the_daemon in the Gradle documentation.
[builder]       Daemon will be stopped at the end of the build
[builder]       > Task :compileJava NO-SOURCE
[builder]       > Task :processResources NO-SOURCE
[builder]       > Task :classes UP-TO-DATE
[builder]       > Task :jar
[builder]       > Task :assemble
[builder]

===> EXPORTING
[exporter] Reusing layer 'paketo-buildpacks/ca-certificates:helper'
[exporter] Reusing layer 'paketo-buildpacks/bellsoft-liberica:helper'
[exporter] Reusing layer 'paketo-buildpacks/bellsoft-liberica:java-security-properties'
[exporter] Reusing layer 'paketo-buildpacks/bellsoft-liberica:jre'
[exporter] Reusing layer 'buildpacksio/lifecycle:launch.sbom'
[exporter] Reusing 1/1 app layer(s)
[exporter] Reusing layer 'buildpacksio/lifecycle:launcher'
[exporter] Reusing layer 'buildpacksio/lifecycle:config'
[exporter] Adding label 'io.buildpacks.lifecycle.metadata'
[exporter] Adding label 'io.buildpacks.build.metadata'
[exporter] Adding label 'io.buildpacks.project.metadata'
[exporter] no default process type
[exporter] Saving imagenew22...
[exporter] *** Images (e95f14c5c820):
[exporter]       imagenew22

When I am trying to execute the image its not getting started because of no process type so I added a procfile with /bin/bash to check whats inside the image but there is nothing under workspace

@dmikusa , any suggestions on this

**Comment by dmikusa** [comment]
The main difference between spring Boot build tools and `pack build` is that the former will build you JAR file locally and then have buildpacks create an image from the JAR file. With pack you can do that, but it's not the default option. To do that with `pack build`, you need to add the `-p target/your-app.jar` argument. You might try that as it'll give you something that's closer to what you get with Spring Boot Build Tools.

The output from your pack build is a little confusing cause it says 11 buildpacks are detecting, but I definitely don't see that many running. It's not clear why from the info included. If you can post the full build output, that would be helpful.

Aside from that, what I can say is that you get the process types from:

- executable-jar
- apache-tomcat
- apache-tomee
- liberty
- dist-zip
- procfile

The spring-boot buildpack does not actually add process types. Most of the time for Spring Boot apps, that happens from executable-jar, because that's the default output format for Spring Boot apps.

Since you're not seeing process types, you'd want to look most closely at the output for the buildpack you expect to contribute the process types and see what it's saying. Also, look closely at Gradle too if you build from source in the build container. This is because Gradle has to detect your output artifact, which is does by looking for certain patterns. It can in some case find the wrong thing, which can lead to confusing issues like this (if it finds nothing, you get an error).

Hope that helps!
